File: BACKEND/src/controllers/authController.ts
import { Request, Response, NextFunction } from 'express';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import crypto from 'crypto';
import { supabaseService } from '../config/supabase';
import { logger } from '../utils/logger';
import { AppError, asyncHandler } from '../middleware/errorHandler';
import { User, UserType, AuthResponse, NotificationPreferences } from '../types';
import { emailService } from '../services/emailService';

interface AuthRequest extends Request {
  user?: any;
}

// Helper function to generate JWT tokens
const generateTokens = (userId: string) => {
  const accessToken = jwt.sign(
    { userId },
    process.env.JWT_SECRET!,
    { expiresIn: process.env.JWT_EXPIRES_IN || '7d' }
  );

  const refreshToken = jwt.sign(
    { userId, type: 'refresh' },
    process.env.JWT_SECRET!,
    { expiresIn: '30d' }
  );

  return { accessToken, refreshToken };
};

// Helper function to determine user type based on document
const determineUserType = (documentType: string, documentNumber: string): UserType => {
  if (documentType === 'NIT') {
    // Could be juridical or company - for now, default to juridical
    // In a real app, you might have additional logic here
    return UserType.JURIDICAL;
  }
  return UserType.NATURAL;
};

export const register = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const {
    email,
    password,
    first_name,
    last_name,
    document_number,
    document_type,
    user_type,
    phone_number,
    company_id
  } = req.body;

  // Check if user already exists
  const existingUsers = await supabaseService.select('users', {
    filters: { email }
  });

  if (existingUsers && existingUsers.length > 0) {
    return next(new AppError('User already exists with this email', 400));
  }

  // Check if document number is already registered
  const existingDocument = await supabaseService.select('users', {
    filters: { document_number }
  });

  if (existingDocument && existingDocument.length > 0) {
    return next(new AppError('User already exists with this document number', 400));
  }

  // Register with Supabase Auth
  const authResult = await supabaseService.signUp(email, password, {
    first_name,
    last_name,
    document_number,
    document_type,
    user_type
  });

  if (!authResult.user) {
    return next(new AppError('Registration failed', 500));
  }

  // Create user profile in database
  const defaultNotificationPrefs: NotificationPreferences = {
    email_enabled: true,
    sms_enabled: false,
    in_app_enabled: true,
    sound_enabled: true,
    process_updates: true,
    hearing_reminders: true,
    document_alerts: true,
    weekly_summary: false
  };

  const userProfile = await supabaseService.insert('users', {
    id: authResult.user.id,
    email,
    first_name,
    last_name,
    document_number,
    document_type,
    user_type: user_type || determineUserType(document_type, document_number),
    phone_number,
    company_id,
    is_active: true,
    email_verified: false,
    notification_preferences: defaultNotificationPrefs,
    created_at: new Date(),
    updated_at: new Date()
  });

  // Generate tokens
  const tokens = generateTokens(authResult.user.id);

  // Send verification email
  try {
    await emailService.sendVerificationEmail(email, authResult.user.id);
  } catch (error) {
    logger.error('Failed to send verification email:', error);
  }

  logger.info(`User registered: ${email}`);

  const response: AuthResponse = {
    user: {
      ...userProfile,
      password_hash: undefined
    },
    access_token: tokens.accessToken,
    refresh_token: tokens.refreshToken,
    expires_in: 7 * 24 * 60 * 60 // 7 days in seconds
  };

  res.status(201).json({
    success: true,
    message: 'Registration successful. Please check your email for verification.',
    data: response
  });
});

export const login = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const { email, password } = req.body;

  // Sign in with Supabase
  const authResult = await supabaseService.signIn(email, password);

  if (!authResult.user) {
    return next(new AppError('Invalid email or password', 401));
  }

  // Get user profile from database
  const userProfile = await supabaseService.select('users', {
    filters: { id: authResult.user.id }
  });

  if (!userProfile || userProfile.length === 0) {
    return next(new AppError('User profile not found', 404));
  }

  const user = userProfile[0];

  // Check if user is active
  if (!user.is_active) {
    return next(new AppError('Account is deactivated', 401));
  }

  // Generate tokens
  const tokens = generateTokens(authResult.user.id);

  // Update last login
  await supabaseService.update('users', user.id, {
    updated_at: new Date()
  });

  logger.info(`User logged in: ${email}`);

  const response: AuthResponse = {
    user: {
      ...user,
      password_hash: undefined
    },
    access_token: tokens.accessToken,
    refresh_token: tokens.refreshToken,
    expires_in: 7 * 24 * 60 * 60 // 7 days in seconds
  };

  res.json({
    success: true,
    message: 'Login successful',
    data: response
  });
});

export const logout = asyncHandler(async (req: AuthRequest, res: Response, next: NextFunction) => {
  const authHeader = req.headers.authorization;
  const token = authHeader && authHeader.split(' ')[1];

  if (token) {
    await supabaseService.signOut(token);
  }

  logger.info(`User logged out: ${req.user?.email}`);

  res.json({
    success: true,
    message: 'Logout successful'
  });
});

export const refreshToken = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const { refresh_token } = req.body;

  if (!refresh_token) {
    return next(new AppError('Refresh token is required', 400));
  }

  try {
    // Refresh session with Supabase
    const result = await supabaseService.refreshSession(refresh_token);

    if (!result.user) {
      return next(new AppError('Invalid refresh token', 401));
    }

    // Generate new tokens
    const tokens = generateTokens(result.user.id);

    // Get user profile
    const userProfile = await supabaseService.select('users', {
      filters: { id: result.user.id }
    });

    const response: AuthResponse = {
      user: {
        ...userProfile[0],
        password_hash: undefined
      },
      access_token: tokens.accessToken,
      refresh_token: tokens.refreshToken,
      expires_in: 7 * 24 * 60 * 60
    };

    res.json({
      success: true,
      message: 'Token refreshed successfully',
      data: response
    });
  } catch (error) {
    return next(new AppError('Invalid refresh token', 401));
  }
});

export const getProfile = asyncHandler(async (req: AuthRequest, res: Response, next: NextFunction) => {
  const user = req.user;

  res.json({
    success: true,
    data: {
      ...user,
      password_hash: undefined
    }
  });
});

export const updateProfile = asyncHandler(async (req: AuthRequest, res: Response, next: NextFunction) => {
  const { first_name, last_name, phone_number, notification_preferences } = req.body;
  const userId = req.user.id;

  const updateData: any = {
    updated_at: new Date()
  };

  if (first_name) updateData.first_name = first_name;
  if (last_name) updateData.last_name = last_name;
  if (phone_number) updateData.phone_number = phone_number;
  if (notification_preferences) updateData.notification_preferences = notification_preferences;

  const updatedUser = await supabaseService.update('users', userId, updateData);

  logger.info(`Profile updated for user: ${req.user.email}`);

  res.json({
    success: true,
    message: 'Profile updated successfully',
    data: {
      ...updatedUser,
      password_hash: undefined
    }
  });
});

export const changePassword = asyncHandler(async (req: AuthRequest, res: Response, next: NextFunction) => {
  const { current_password, new_password } = req.body;
  const userId = req.user.id;

  // This would typically involve Supabase's password update functionality
  // For now, we'll log the attempt
  logger.info(`Password change requested for user: ${req.user.email}`);

  res.json({
    success: true,
    message: 'Password changed successfully'
  });
});

export const forgotPassword = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const { email } = req.body;

  // Check if user exists
  const users = await supabaseService.select('users', {
    filters: { email }
  });

  if (!users || users.length === 0) {
    return next(new AppError('No user found with this email', 404));
  }

  // Generate reset token
  const resetToken = crypto.randomBytes(32).toString('hex');
  const resetExpires = new Date(Date.now() + 10 * 60 * 1000); // 10 minutes

  // Store reset token (you'd need a password_resets table)
  // For now, just log
  logger.info(`Password reset requested for: ${email}`);

  try {
    await emailService.sendPasswordResetEmail(email, resetToken);
  } catch (error) {
    logger.error('Failed to send password reset email:', error);
    return next(new AppError('Failed to send reset email', 500));
  }

  res.json({
    success: true,
    message: 'Password reset email sent'
  });
});

export const resetPassword = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const { token, new_password } = req.body;

  // Verify reset token (implementation depends on your reset token storage)
  // For now, just log
  logger.info(`Password reset attempted with token: ${token}`);

  res.json({
    success: true,
    message: 'Password reset successful'
  });
});

export const verifyEmail = asyncHandler(async (req: Request, res: Response, next: NextFunction) => {
  const { token } = req.params;

  // Verify email token (implementation depends on your token storage)
  // For now, just log
  logger.info(`Email verification attempted with token: ${token}`);

  res.json({
    success: true,
    message: 'Email verified successfully'
  });
});

File: SPRING-BACKEND/src/main/java/com/judicial/processes/controller/AuthController.java
package com.judicial.processes.controller;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.judicial.processes.dto.ApiResponse;
import com.judicial.processes.dto.AuthDTO.*;
import com.judicial.processes.dto.AuthDTO.AuthResponse;
import com.judicial.processes.dto.AuthDTO.ChangePasswordRequest;
import com.judicial.processes.dto.AuthDTO.ForgotPasswordRequest;
import com.judicial.processes.dto.AuthDTO.LoginRequest;
import com.judicial.processes.dto.AuthDTO.RefreshTokenRequest;
import com.judicial.processes.dto.AuthDTO.RegisterRequest;
import com.judicial.processes.dto.AuthDTO.ResetPasswordRequest;
import com.judicial.processes.dto.AuthDTO.UserDTO;
import com.judicial.processes.service.AuthService;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/auth")
@CrossOrigin(origins = {"http://localhost:3000", "http://localhost:3001", "http://localhost:19006"})
public class AuthController {
    
    private static final Logger logger = LoggerFactory.getLogger(AuthController.class);
    
    @Autowired
    private AuthService authService;
    
    @PostMapping("/register")
    public ResponseEntity<ApiResponse<AuthResponse>> register(
            @Valid @RequestBody RegisterRequest request,
            HttpServletRequest httpRequest) {
        try {
            logger.info("Registration attempt for email: {}", request.getEmail());
            
            AuthResponse authResponse = authService.register(request);
            
            logger.info("User registered successfully: {}", request.getEmail());
            
            return ResponseEntity.status(201).body(
                new ApiResponse<>(true, authResponse, 
                "Registration successful. Please check your email for verification.")
            );
            
        } catch (RuntimeException e) {
            logger.error("Registration failed for email: " + request.getEmail(), e);
            return ResponseEntity.badRequest().body(
                new ApiResponse<>(false, e.getMessage())
            );
        } catch (Exception e) {
            logger.error("Unexpected registration error for email: " + request.getEmail(), e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Registration failed")
            );
        }
    }
    
    @PostMapping("/login")
    public ResponseEntity<ApiResponse<AuthResponse>> login(
            @Valid @RequestBody LoginRequest request,
            HttpServletRequest httpRequest) {
        try {
            logger.info("Login attempt for email: {}", request.getEmail());
            
            AuthResponse authResponse = authService.login(request.getEmail(), request.getPassword());
            
            if (authResponse != null) {
                logger.info("User logged in successfully: {}", request.getEmail());
                return ResponseEntity.ok(
                    new ApiResponse<>(true, authResponse, "Login successful")
                );
            } else {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Invalid email or password")
                );
            }
            
        } catch (RuntimeException e) {
            logger.error("Login failed for email: " + request.getEmail(), e);
            return ResponseEntity.status(401).body(
                new ApiResponse<>(false, e.getMessage())
            );
        } catch (Exception e) {
            logger.error("Unexpected login error for email: " + request.getEmail(), e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Login failed")
            );
        }
    }
    
    @PostMapping("/logout")
    public ResponseEntity<ApiResponse<Void>> logout(HttpServletRequest request) {
        try {
            String authHeader = request.getHeader("Authorization");
            String token = null;
            
            if (authHeader != null && authHeader.startsWith("Bearer ")) {
                token = authHeader.substring(7);
            }
            
            if (token != null) {
                authService.logout(token);
            }
            
            logger.info("User logged out successfully");
            
            return ResponseEntity.ok(
                new ApiResponse<>(true, "Logout successful")
            );
            
        } catch (Exception e) {
            logger.error("Logout error", e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Logout failed")
            );
        }
    }
    
    @PostMapping("/refresh-token")
    public ResponseEntity<ApiResponse<AuthResponse>> refreshToken(
            @Valid @RequestBody RefreshTokenRequest request) {
        try {
            AuthResponse authResponse = authService.refreshToken(request.getRefreshToken());
            
            if (authResponse != null) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, authResponse, "Token refreshed successfully")
                );
            } else {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Invalid refresh token")
                );
            }
            
        } catch (Exception e) {
            logger.error("Token refresh error", e);
            return ResponseEntity.status(401).body(
                new ApiResponse<>(false, "Invalid refresh token")
            );
        }
    }
    
    @GetMapping("/profile")
    public ResponseEntity<ApiResponse<UserDTO>> getProfile(HttpServletRequest request) {
        try {
            String authHeader = request.getHeader("Authorization");
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Access token is required")
                );
            }
            
            String token = authHeader.substring(7);
            UserDTO user = authService.getProfile(token);
            
            if (user != null) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, user, null)
                );
            } else {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Invalid token")
                );
            }
            
        } catch (Exception e) {
            logger.error("Get profile error", e);
            return ResponseEntity.status(401).body(
                new ApiResponse<>(false, "Authentication failed")
            );
        }
    }
    
    @PutMapping("/profile")
    public ResponseEntity<ApiResponse<UserDTO>> updateProfile(
            @RequestBody UserDTO updateData,
            HttpServletRequest request) {
        try {
            String authHeader = request.getHeader("Authorization");
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Access token is required")
                );
            }
            
            String token = authHeader.substring(7);
            UserDTO updatedUser = authService.updateProfile(token, updateData);
            
            if (updatedUser != null) {
                logger.info("Profile updated successfully");
                return ResponseEntity.ok(
                    new ApiResponse<>(true, updatedUser, "Profile updated successfully")
                );
            } else {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Invalid token")
                );
            }
            
        } catch (Exception e) {
            logger.error("Profile update error", e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Profile update failed")
            );
        }
    }
    
    @PostMapping("/change-password")
    public ResponseEntity<ApiResponse<Void>> changePassword(
            @Valid @RequestBody ChangePasswordRequest request,
            HttpServletRequest httpRequest) {
        try {
            String authHeader = httpRequest.getHeader("Authorization");
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return ResponseEntity.status(401).body(
                    new ApiResponse<>(false, "Access token is required")
                );
            }
            
            String token = authHeader.substring(7);
            boolean success = authService.changePassword(token, request.getCurrentPassword(), request.getNewPassword());
            
            if (success) {
                logger.info("Password changed successfully");
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Password changed successfully")
                );
            } else {
                return ResponseEntity.status(400).body(
                    new ApiResponse<>(false, "Current password is incorrect")
                );
            }
            
        } catch (Exception e) {
            logger.error("Change password error", e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Password change failed")
            );
        }
    }
    
    @PostMapping("/forgot-password")
    public ResponseEntity<ApiResponse<Void>> forgotPassword(
            @Valid @RequestBody ForgotPasswordRequest request) {
        try {
            boolean success = authService.forgotPassword(request.getEmail());
            
            if (success) {
                logger.info("Password reset email sent to: {}", request.getEmail());
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Password reset email sent successfully")
                );
            } else {
                return ResponseEntity.status(404).body(
                    new ApiResponse<>(false, "No user found with this email")
                );
            }
            
        } catch (Exception e) {
            logger.error("Forgot password error for email: " + request.getEmail(), e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Failed to send reset email")
            );
        }
    }
    
    @PostMapping("/reset-password")
    public ResponseEntity<ApiResponse<Void>> resetPassword(
            @Valid @RequestBody ResetPasswordRequest request) {
        try {
            boolean success = authService.resetPassword(request.getToken(), request.getNewPassword());
            
            if (success) {
                logger.info("Password reset successful");
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Password reset successful")
                );
            } else {
                return ResponseEntity.status(400).body(
                    new ApiResponse<>(false, "Invalid or expired reset token")
                );
            }
            
        } catch (Exception e) {
            logger.error("Password reset error", e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Failed to reset password")
            );
        }
    }
    
    @GetMapping("/verify-email/{token}")
    public ResponseEntity<ApiResponse<Void>> verifyEmail(@PathVariable String token) {
        try {
            boolean success = authService.verifyEmail(token);
            
            if (success) {
                logger.info("Email verification successful for token: {}", token);
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Email verified successfully")
                );
            } else {
                return ResponseEntity.status(400).body(
                    new ApiResponse<>(false, "Invalid or expired verification token")
                );
            }
            
        } catch (Exception e) {
            logger.error("Email verification error for token: " + token, e);
            return ResponseEntity.internalServerError().body(
                new ApiResponse<>(false, "Email verification failed")
            );
        }
    }
}